### PR TITLE
Fix Bug `Uncreate()` with `rate_func` via introducing new parameter `reversed` to class `Animation` 

### LINKED ISSUE #2469: `Uncreate` does not work as expected when overriding the default argument for `rate_func`
## Description of bug / unexpected behavior
When passing any argument for `rate_func` to the `Uncreate` animation, the result is exactly the same as would be seen from using the `Create` animation. For instance,

```py
class Bug(Scene):
    def construct(self):
      s = Square()
      self.add(s)
      self.play(Uncreate(s), rate_func=smooth)
      self.wait()
```

produces:

https://user-images.githubusercontent.com/9102529/150136839-6073c3be-01cb-4bf7-8931-3968a468aa49.mp4

Currently, `Uncreate` just inherits from `Create` and overrides the default `rate_func` argument. This is a poor implementation; it should either inherit directly from `ShowPartial`, or a general method to reverse animations should be implemented. The `Uncreate` animation will then just be a special case of a reversed animation.

## How to reproduce the issue

<details><summary>Code for reproducing the problem</summary>

```py
Paste your code here.
```

</details>

## Additional media files

<details><summary>Images/GIFs</summary>

</details>

## Logs
<details><summary>Terminal output</summary>

```
PASTE HERE OR PROVIDE LINK TO https://pastebin.com/ OR SIMILAR
```

</details>

## System specifications

<details><summary>System Details</summary>

- OS (with version, e.g Windows 10 v2004 or macOS 10.15 (Catalina)):
- RAM:
- Python version (`python/py/python3 --version`):
- Installed modules (provide output from `pip list`):
```
PASTE HERE
```
</details>

<details><summary>LaTeX details</summary>

+ LaTeX distribution (e.g. TeX Live 2020):
+ Installed LaTeX packages:

</details>

<details><summary>FFMPEG</summary>

Output of `ffmpeg -version`:

```
PASTE HERE
```
</details>

### PR DESCRIPTION
**Pull request by darkway-s**
## Overview: What does this pull request change?

Fix #2469. 
Refractor the `Uncreate()` in manim/animation/creation.py. The new implementation uses a flag member `reversed`. Set it to true and its super class handles the reverse.

Introduce a bool parameter `reversed` to class `Animation` in manim/animation/animation.py. It decides whether the animation need to be played backwards. Default to be False.
Add conditional branches in `get_sub_alpha()` of class `Animation`. If the parameter `reversed` is True, it would set `rate_func(t)` to `rate_func(1 - t)`. 

Add a unit test in tests/test_graphical_units/test_creation.py to cover this branch.

Here is an example to use it.

```python
class UncreateSquare(Scene):
    def construct(self):
        square = Square()
        self.add(square)
        self.play(Uncreate(square), rate_func=linear)
```

## Motivation and Explanation: Why and how do your changes improve the library?

It's really weird to call the function `Uncreate()` with `rate_func=lambda t, inflection=10.0: smooth((1-t), inflection`. It results from the hard-code implementation.
So we made a bit refactoring on `Uncreate()` and now it works more smoothly. It's more convenient for users to use it.
## Reviewer Checklist
- [ ] The PR title is descriptive enough for the changelog, and the PR is labeled correctly
- [ ] If applicable: newly added non-private functions and classes have a docstring including a short summary and a PARAMETERS section
- [ ] If applicable: newly added functions and classes are tested

### CODE REVIEW

**Review by behackl — COMMENTED** [review]
Hello! Thank you for your contribution!

It definitely is good that this is fixed, but I am not sure whether I like the solution of introducing yet another keyword argument (I do appreciate the added documentation though!).

What I don't like about this is that it introduces yet another way to solve the same problem: in theory, `FadeOut`, `Unwrite` and `Uncreate` could be handled similarly. In practice there are now three completely separate approaches for dealing with them.

In general I think this can stay; here are two things to consider:
- The implementation of `Unwrite` should be changed to use `reverse` as well.
- Should `reverse` be made smart, in the sense that if an remover animation is reversed it gets flagged as an introducer, and introducers get flagged as removers?

**Comment by darkway-s** [comment]
Thanks for your reviewing!

Actually I use the same solution as `Unwrite` to implement `Uncreate`, inverting the parameter `rate_func`. In the implementation of `Unwrite`, it simply inverts it in the function itself. 

I feel like we have done some duplication of work on these things. So I put this functionality into their public superclass, `Animation`. I would be more convenient for other developers to use.

I will change the implementation of `Unwirte` to use `reverse` as well. 



I fail to understand how `FadeOut` works. It seems that it only enables `remover`. 

I would appreciate a lot if you can tell me how it works.

**Comment by behackl** [comment]
It's fine if we keep `FadeOut` the way it is for now. I'm happy if `Unwrite` is changed to the same logic. 👍 

The question regarding the `is_remover` and `is_introducer` keyword arguments still remains. I am somewhat torn whether `Animation(reversed=True, introducer=True)` should actually set `is_remover` to `True` (and vice versa with `remover=True` and `is_introducer`).

**Comment by darkway-s** [comment]
It is a question if the condition exists that `reversed` is true but `remover` is false. As far as the current usages is concerned, `remover=True` is a necessary condition of `reversed=True`.

**Comment by darkway-s** [comment]
The inside of the `Unwrite` is confusing.
The `Unwrite` also have a parameter called `reverse`, which decides whether the animation starts erasing from the last submobject, while `reversed` is a member of animation that decides the way of playing on one submobject. It won't be confusing because `reversed` is invisible for users. But problems would occur if a user wants to have a UnwriteReverseTrue but wrongly writes `reverse ` to `reversed`. The user won't get the expected result.

**Comment by Sefik-Palazoglu** [comment]
I thought about the problem of making `reversed` parameter smart to reverse the remover and introducer fields. What we had in mind initially when working on this problem was to provide the developers a way to implement Uncreate() and similar 'reversed' functions while actually reversing them (not just by changing the corresponding rate_func(t) to rate_func(1-t), because it results in the problem in this issue). 
As a result, what we envisioned was that this parameter `reversed` should be invisible/hidden from the end user but useful when a developer implements a 'reverse' Animation. As it provides a somewhat easy and uniform way to reverse a given animation.
My main point against the smart implementation is this, if we let the end user able to manually determine the behavior of the animation through the new kwarg `reversed`, I think we have no way of knowing the intention of the end-user whether they wanted to make a remover or introducer animation.
Also, there are animation which are both not removers and introducers, certainly these should not be both removers and introducers when reversed.
So, I don't yet have a solution about the `is_remover` and `is_introducer` parameters yet, but believe the correct approach is to hide this functionality from the end-user.

TL;DR: `reversed` is intended to be utilized by developers, burden of assigning correct values to `is_remover` and `is_introducer` is on the developer of the 'reversed' animation. End users are not intended to change the `reversed` parameter. Handling the smart request will probably lead to more edge-cases than we are trying to solve currently.

**Comment by Sefik-Palazoglu** [comment]
Also, a somewhat similar problem is: What happens if a user tries something like `self.play(Unwrite(), introducer=true)`? I think in this case, this animation becomes both an introducer and remover. Is this currently handled or is it considered a misuse?

**Comment by behackl** [comment]
Thank you for your elaborate thoughts! I agree, making `reverse` smart in the sense that remover and introducer settings would get switched likely leads to more confusion than necessary. However, there are no "developer-only" keyword arguments, and I am against introducing a concept like that.

There is a simple solution to better capture the intention / semantics introduced with the new keyword argument; here is what I propose:
- Rename `reversed` to the more explicit `reverse_rate_function`,
- And explicitly mention the fact that passing `reverse_rate_function=True` to animations that introduce or remove mobjects also requires the corresponding changes to the `remover` and `introducer` keyword arguments.

What do you two think about that, @darkway-s @Sefik-Palazoglu?

**Comment by darkway-s** [comment]
LGTM!

**Comment by Sefik-Palazoglu** [comment]
I agree with your proposition, looks good!

File: manim/animation/animation.py [hunk 1]
```diff
@@ -49,6 +49,13 @@ class Animation:
 
         For example ``rate_func(0.5)`` is the proportion of the animation that is done
         after half of the animations run time.
+
+
+    reverse_rate_function
+        Whether the animation need to be played backwards.
+
+        Setting `reverse_rate_function` doesn't have any effect on `remover` and `introducer`.
+            If you want your animation to be a introducer or remover. You must set it manually.
```
Comment (behackl) [inline manim/animation/animation.py:58]: Minor suggestion regarding wording.

```suggestion
        Reverses the rate function of the animation. Setting ``reverse_rate_function``
        does not have any effect on ``remover`` or ``introducer``. These need to be
        set explicitly if an introducer-animation should be turned into a remover one
        and vice versa.
```

File: manim/animation/creation.py [hunk 2]
```diff
@@ -403,7 +403,7 @@ def __init__(
             vmobject,
             run_time=run_time,
             lag_ratio=lag_ratio,
-            rate_func=lambda t: -rate_func(t) + 1,
+            reverse_rate_function=True,  # `reverse_rate_function` is in the kwards of class `Animation`, which decides whether the animation need to be played backwards.
```
Comment (behackl) [inline manim/animation/creation.py:406]: Not sure whether I understand this comment, it could probably be removed.

**Review by behackl — CHANGES_REQUESTED** [review]
LGTM in general, two more minor remarks.

File: manim/animation/animation.py (hunk 1 above)
Comment (darkway-s) [reply manim/animation/animation.py:58]: LGTM!

File: manim/animation/creation.py (hunk 2 above)
Comment (darkway-s) [reply manim/animation/creation.py:406]: I added the comment because the name of the parameter `reversed` or `reverse_rate_function` is similar to `reverse`, which is a also parameter, but defined in `Write()`. With this comment, it could be better understood for developers when they look into `Write()`

Comment (darkway-s) [reply manim/animation/creation.py:406]: I think it's OK to remove it after we have assign such a explicit name to the parameter.

**Review by behackl — APPROVED** [review]
